rust/protocol/src/sealed_sender.rs

```diff
@@ -879,8 +1031,133 @@ mod sealed_sender_v2 {
 
         HKDF::new(3)?.derive_secrets(&agreement_key_input, LABEL_DH_S, AUTH_TAG_LEN)
     }
+
+    #[test]
+    #[allow(clippy::unwrap_used)]
+    fn test_agreement_and_authentication() -> Result<()> {
+        // The sender and recipient each have a long-term identity key pair.
+        let sender_keypair = KeyPair::generate(&mut rand::thread_rng());
+        let sender_identity = IdentityKeyPair::try_from(sender_keypair).unwrap();
+
+        let recipient_keypair = KeyPair::generate(&mut rand::thread_rng());
+        let recipient_identity = IdentityKeyPair::try_from(recipient_keypair).unwrap();
+
+        // Generate random bytes used for our multi-recipient encoding scheme.
+        let m: [u8; MESSAGE_KEY_LEN] = (&mut rand::thread_rng()).gen();
+        // Derive an ephemeral key pair from those random bytes.
+        let ephemeral_keys = DerivedKeys::calculate(&m);
+        let ephemeral_public_key = ephemeral_keys.e.public_key;
+
+        // Encrypt the ephemeral key pair.
+        let sender_c_0 = apply_agreement_xor(
+            &ephemeral_keys.e,
+            &recipient_keypair.public_key,
+            Direction::Sending,
+            &m,
+        )?;
+        // Compute an authentication tag for the encrypted key pair.
+        let sender_at_0 = compute_authentication_tag(
+            &sender_identity,
+            recipient_identity.identity_key(),
+            Direction::Sending,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+
+        // The message recipient calculates the original random bytes and authenticates the result.
+        let recv_m = apply_agreement_xor(
+            &recipient_keypair,
+            &ephemeral_public_key,
+            Direction::Receiving,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_m.as_ref(), m.as_ref());
+
+        let recv_at_0 = compute_authentication_tag(
+            &recipient_identity,
+            sender_identity.identity_key(),
+            Direction::Receiving,
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_at_0.as_ref(), sender_at_0.as_ref());
+
+        Ok(())
+    }
 }
 
+/// This method implements a single-key multi-recipient [KEM], defined in Manuel Barbosa's
```

Comment:
"as defined", I think, since we're using a particular set of parameters for Barbosa's paper.

Reply:
Thanks! Changed!